### ISSUE TITLE
request for comments: Remove object.options? Add Object.assign() polyfill?

### ISSUE THREAD

**Issue by jpommerening** [LaxarJS/laxar#402]
@alex3683, @x1B:

### Should we remove `object.options()`?

It is straightforward to implement oneself. I *think* using the spread operator is clearer (no confusion regarding parameter order and count), but that may be just my personal preference.

### Should we polyfill `Object.assign()`?

With some caveats `Object.assign` can be implemented using the spread operator:
```js
target = {
   ...target,
   ...source1,
   ...source2
};
```
However, not only is this inefficient, because it creates a new object and unnecessarily iterates over the keys of the target object. It also does not update the target object itself, but only the reference in the current scope, while other refs remain unchanged.

Here's what Babel generates:
```js
"use strict";var _extends = Object.assign || function (target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i];for (var key in source) {if (Object.prototype.hasOwnProperty.call(source, key)) {target[key] = source[key];}}}return target;};

var target = _extends({}, target, source);
```

The cases where you really want to modify an existing map are few and far between. After a very quick search I only found two locations that could benefit from `Object.assign()`:
- [Assignment to `env.defaultOptions` in `json_validator.js`](https://github.com/LaxarJS/laxar/blob/04259d6287deec851d2536c6074efe88530c984b/lib/utilities/json_validator.js#L59)
- [Assignment to `widgetAdapters` in `widget_loader.js`](https://github.com/LaxarJS/laxar/blob/480010db78624a60cd2cf1d8a4359c953d628947/lib/loaders/widget_loader.js#L134) (wip of #396)

Since these are only hit once in the usual application lifecycle, I'm not sure the tradeoff is worthwhile. There's another point to this however. Newer versions of [babel-plugins-transform-object-rest-spread](https://babeljs.io/docs/plugins/transform-object-rest-spread/) have an option called `useBuiltIns` to not generate the `_extend()` function and instead generate code using `Object.assign()`. Since the `_extend()` function is generated once per module, this could shave a few bytes off our bundle:

```console
# before
$ wc dist/laxar.min.js dist/polyfills.js
       3    2094   69728 dist/laxar.min.js
    1701    5915   53220 dist/polyfills.js
    1704    8009  122948 total
# after
$ wc dist/laxar.min.js dist/polyfills.js
       3    2036   68428 dist/laxar.min.js
    1715    5957   53611 dist/polyfills.js
    1718    7993  122039 total
```

**Comment by x1B** [comment]
@jpommerening I suggest keeping `object.options` to simplify migration to v2.0, as it is one of the few library methods that are actually used quite often by widgets and libraries.

Under the hood, I think we can go for the `assign` polyfill, which will reduce the cost for keeping `object.options` to a negligible amount.

**Comment by jpommerening** [comment]
Okay, that's reasonable.

Another thing: There's an `object.assign` polyfill module on npm. It weighs about 12k (6k minified), and that is only because it shares some dependencies with `array.from`. Together they come in at about 45k (20k minified), putting the bundled polyfills close to 67k (33.5k minified). The implementations on MDN are 4k (1.2k minified) together, resulting in a polyfill bundle of 26k (15k minified).

Modules and code reuse are nice and all, but the browsers that require polyfills often suffer the most from large downloads.

@alex3683, @x1B: What's your opinion on this?

**Comment by x1B** [comment]
Personally, I'd be fine with the MDN implementation. Let's keep that file size down. 33k minimized seems way too large for this functionality.

**Comment by jpommerening** [comment]
Implemented on `master` (`v2.0.0`)

### Summary

We're now bringing our own polyfill implementations for `Array.from()`, `Array.includes()` and `Object.assign()` which are based on those from MDN. They live in separate files inside `lib/polyfills` and are imported by the `polyfills.js` entry point.

The Babel plugin `transform-object-rest-spread` is now configured to generate `Object.assign()` calls.

Oh and I removed the `retainLines` option from `.babelrc`, since the Babel output is streamed to webpack anyway and the (non-minified) webpack bundle is more readable that way. This actually slipped in accidentally but I believe it won't cause any harm.

**Comment by x1B** [comment]
Sounds great :+1:
